kv/dist_sender.go

```diff
@@ -464,16 +464,24 @@ func (ds *DistSender) sendRPC(desc *proto.RangeDescriptor,
 		a.Header().Replica = *replicaMap[addr.String()]
 		return a
 	}
-	firstReply := true
+	// RPC send will use go routine to send request and fill reply, and there is
+	// no lock for reply varible, so it may have race case which just merged
+	// reply got overwritten by RPC returned value if reply itself as RPC parameter
+	// in first try when there are multiple address to try.
+	// So here reply itself is not used as RPC parameter if there is more
+	// than 1 address to try.
 	getReply := func() interface{} {
-		if firstReply {
-			firstReply = false
+		if len(addrs) == 1 {
```

Comment:
This means that in practice (for any nontrivial single node test) we'll always clone the reply. That's unfortunate, but you're also correct in that there's a data race in the way the reply was used prior to this change. Also, `reply` will be an empty protobuf, so it's not a large structure. I would even avoid special-casing the case `len(addrs) == 1` because it's of no importance to production setups, so all it can do is skew single-machine benchmarks.